DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+{
+    A a;
+    fun(a);             // calls 1 - conversion to const match; standard copying
+
+    const A a1;
+    fun(a1);             // calls 1 - direct match; standard copying
+
+    shared A a2;
+    fun(a2);             // calls 2 - conversion with copy ctor match; call copy ctor on parameter
+
+    B b;
+    fun(b);              // error: both gun functions match, because copy
+                         // ctor matches both ctor1 and ctor2
+
+    const B b1;
+    fun(b1)               // calls 4 - direct match; standard copying
+
+    immutable B b2;
+    fun(b2);              // calls 4 - conversion to const match; call ctor4
+
+    shared B b3;
+    fun(b3);              // calls 4 - direct match; call ctor3
+}
+```
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref B another) immutable {}
+
+}
+
+void main()
+{
+    immutable A ia;
+    A a = ia;            // 1 - calls copy constructor
+
+    B bc;
+    B b = bc;            // 2 - b is evaluated to B.fun
+}
+```
+
+In situations where both the copy constructor and `alias this` are suitable
+to solve the assignment (1), the copy constructor takes precedence over `alias this`
+because it is considered more specialized (the sole purpose of the copy constructor is to
+create copies). However, if no copy constructor in the overload set matches the exact
+qualified types of the source and the destination, the `alias this` is preferred (2).
+
+#### Generating copy constructors
+
+A copy constructor is generated for a `struct` S  if one of S's fields defines
+a copy constructor that S does not define.
+
+A field of a `struct` may define multiple copy constructors. In this situation,
+a copy constructor is generated for each overload.
+
+```d
+struct M
+{
+    @implicit this(ref immutable M) {}
+    @implicit this(ref immutable M) shared {}
+}
+
+struct S
+{
+    M m;
+    @implicit this(ref shared S) {}
+    /* The copy constructors with the following signatures are generated:
+        @implicit this(ref immutable S)
+        @implicit this(ref immutable S) shared
+     */
+}
+```
+
+The body of all the generated copy constructors does memberwise initialization for all
+the fields:
+
+```d
+{
+    this.field1 = s.field1;
+    this.field2 = s.field2;
+    ...;
+}
+```
+
+Inside the body of the generated copy constructors, for the fields that define a copy
+constructor, the assignment will be rewritten to a call to it; for those that do not,
+standard copying is employed, if possible. If certain fields cannot perform copies
```

Comment:
The `alias this` caveat from above throws a wrench in to this as well no? i.e. for those that do not, standard copying is employed, if possible, unless an alias this assignment is possible

Reply:
Indeed.